library/std/src/sync/mpsc/oneshot.rs

```diff
@@ -27,15 +27,15 @@ pub use self::UpgradeResult::*;
 
 use crate::cell::UnsafeCell;
 use crate::ptr;
-use crate::sync::atomic::{AtomicUsize, Ordering};
+use crate::sync::atomic::{AtomicPtr, Ordering};
 use crate::sync::mpsc::blocking::{self, SignalToken};
 use crate::sync::mpsc::Receiver;
 use crate::time::Instant;
 
 // Various states you can find a port in.
-const EMPTY: usize = 0; // initial state: no data, no blocked receiver
-const DATA: usize = 1; // data ready for receiver to take
-const DISCONNECTED: usize = 2; // channel is disconnected OR upgraded
+const EMPTY: *mut u8 = ptr::null_mut(); // initial state: no data, no blocked receiver
+const DATA: *mut u8 = ptr::null_mut::<u8>().wrapping_add(1); // data ready for receiver to take
+const DISCONNECTED: *mut u8 = ptr::null_mut::<u8>().wrapping_add(2); // channel is disconnected OR upgraded
```

Comment:
Maybe just use https://doc.rust-lang.org/nightly/std/ptr/fn.invalid_mut.html instead of the `wrapping_add`s?

For example,
```suggestion
const DISCONNECTED: *mut u8 = ptr::invalid_mut::<u8>(2); // channel is disconnected OR upgraded
```

(I don't think it makes a semantic difference, but it might be a bit simpler.)

Reply:
`2 as *mut u8` should also work.